src/uucore/src/lib/features/format/num_parser.rs

```diff
@@ -0,0 +1,359 @@
+        if self.negative {
+            -n
+        } else {
+            n
+        }
+    }
+
+    /// Parse a number as f64
+    pub fn parse_f64(input: &str) -> Result<f64, ParseError<'_, f64>> {
+        match Self::parse(input, false) {
+            Ok(v) => Ok(v.into_f64()),
+            Err(ParseError::NotNumeric) => Self::parse_f64_special_values(input),
+            Err(e) => Err(e.map(|v, rest| ParseError::PartialMatch(v.into_f64(), rest))),
+        }
+    }
+
+    fn parse_f64_special_values(input: &str) -> Result<f64, ParseError<'_, f64>> {
+        let (sign, rest) = if let Some(input) = input.strip_prefix('-') {
+            (-1.0, input)
+        } else {
+            (1.0, input)
+        };
+        let prefix = rest
+            .chars()
+            .take(3)
+            .map(|c| c.to_ascii_lowercase())
+            .collect::<String>();
+        let special = match prefix.as_str() {
+            "inf" => f64::NEG_INFINITY,
+            "nan" => f64::NAN,
+            _ => return Err(ParseError::NotNumeric),
+        }
+        .copysign(sign);
+        if rest.len() == 3 {
+            Ok(special)
+        } else {
+            Err(ParseError::PartialMatch(special, &rest[3..]))
+        }
+    }
+
+    #[allow(clippy::cognitive_complexity)]
+    fn parse(input: &str, integral_only: bool) -> Result<Self, ParseError<'_, Self>> {
+        if let Some(rest) = input.strip_prefix('\'') {
+            let mut chars = rest.char_indices().fuse();
+            let v = chars.next().map(|(_, c)| Self {
+                base: Base::Decimal,
+                negative: false,
+                integral: u64::from(c),
+                fractional: 0,
+                precision: 0,
+            });
+            return match (v, chars.next()) {
+                (Some(v), None) => Ok(v),
+                (Some(v), Some((i, _))) => Err(ParseError::PartialMatch(v, &rest[i..])),
+                (None, _) => Err(ParseError::NotNumeric),
+            };
+        }
+        let (negative, unsigned) = if let Some(input) = input.strip_prefix('-') {
+            (true, input)
+        } else {
+            (false, input)
+        };
+        let (base, rest) = if let Some(rest) = unsigned.strip_prefix('0') {
+            if let Some(rest) = rest.strip_prefix(['b', 'B']) {
+                (Base::Binary, rest)
+            } else if let Some(rest) = rest.strip_prefix(['x', 'X']) {
+                (Base::Hexadecimal, rest)
+            } else if integral_only {
+                (Base::Octal, unsigned)
+            } else {
+                (Base::Decimal, unsigned)
+            }
+        } else {
+            (Base::Decimal, unsigned)
+        };
+        if rest.is_empty() {
+            return Err(ParseError::NotNumeric);
+        }
+        let (mut index, mut int, mut point, mut frac, mut prec, mut ended) =
+            (0, 0u64, false, 0u64, 0, false);
+        for c in rest.chars() {
```

Comment:
A couple of things about this loop:
- It might be easier to use bytes instead of chars because we only care about the digits which are in ascii range.
- I think it might be nicer to split this loop into before and after the integral part, so encountering a `.` will break the first loop and will go to the second. The current version is a bit hard to follow with all the mutable variables that keep track of the state (though I can't find any problems with it :) )

Reply:
About chars / bytes : sure, but we might encounter a non-ASCII char in the process and we have to deal with it atomically.

About separating the loops, let me check if it makes things simpler.

Thanks for the review!

Reply:
> but we might encounter a non-ASCII

I think we won't split on that if we encounter it, because it won't be a digit or a point or something else we can use. I think it should be safe.

Reply:
To expand on this a bit. The coreutils are weird regarding utf8. The GNU versions usually do not care at all whether a string is valid. The "correct" (i.e. compatible) solution is usually to use bytes throughout and avoid using `str/String` altogether. Then for printing in error messages we do a `to_string_lossy` or something like that. It's a bit messy unfortunately.

For example, I bet that `printf` will happily accept something that starts with digits and then has some invalid utf8. I can trigger this in `fish`:
```
# GNU
> printf %d 10\xf0
printf: ‘10\360’: value not completely converted
10
# uutils
> printf %d 10\xf0
error: invalid UTF-8 was detected in one or more arguments

Usage: target/debug/coreutils printf FORMATSTRING [ARGUMENT]...
       target/debug/coreutils printf FORMAT [ARGUMENT]...
       target/debug/coreutils printf OPTION

Reply:
@tertsdiepraam Breaking the loop in two parts has indeed decreased the complexity a lot. It also made a better structure to add inline comments. Thanks for the suggestion!

Reply:
> To expand on this a bit. The coreutils are weird regarding utf8. The GNU versions usually do not care at all whether a string is valid. The "correct" (i.e. compatible) solution is usually to use bytes throughout and avoid using `str/String` altogether. Then for printing in error messages we do a `to_string_lossy` or something like that. It's a bit messy unfortunately.

Agreed, but it comes back to the previous question: do we want to be compatible with the GNU version even when we go out of the scope of its intended usage, or is it enough to be compatible with the GNU version for its intended usage, and then to be more correct outside of this usage?

Anyways, as far as this parser is concerned, given that it receives a `&str` as input, I suggest to keep `char` which is more readable and should infer to performance penalty, and switch to `u8` if at some point you decide to manipulate `&[u8]` inside of `printf` for example. The change would be easy and feel natural then.

Reply:
> do we want to be compatible with the GNU version even when we go out of the scope of its intended usage, or is it enough to be compatible with the GNU version for its intended usage, and then to be more correct outside of this usage?

So far, the answer has been that we do indeed want to be compatible, but of course the intended usage has more priority. It's difficult to tell what even is the intended usage at this point, because the coreutils have accumulated so much functionality. I think the strength of this particular project lies in the compatibility, even though I often wish I could change (i.e. "fix") the behaviour. But I'm always happy to discuss specific cases :)